### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by phlmn** [shapehq/tartelet#39]
Hi, I'm experiencing extensive memory usage of the Tartelet process. Tartelet runs for less than one day and then does not process jobs anymore, due to the system being out of memory. 

I've configured Tartelet to run two VMs.

![Screenshot 2023-07-06 at 10 44 02](https://github.com/shapehq/tartelet/assets/806995/f9db3bf8-63be-4a4b-a758-faedeb88ec1c)
![Screenshot 2023-07-06 at 10 45 09](https://github.com/shapehq/tartelet/assets/806995/e2a28ba5-f2d6-48de-b760-12289fada60b)
![Screenshot 2023-07-06 at 10 44 20](https://github.com/shapehq/tartelet/assets/806995/fe27b726-ff49-4c9c-8596-ecbc3dcf6488)

**Comment by simonbs** [comment]
@phlmn Are you running the latest version of Tartelet?

I've just checked our runners and I'm not seeing nearly as high a memory usage. See below for a screenshot of Activity Monitor from one of the runners.

<img width="858" alt="Screenshot 2023-07-17 at 14 47 58" src="https://github.com/shapehq/tartelet/assets/830995/4c6dd591-630e-485e-accf-93497e4f9f99">

I also ran the Leaks profiling tool on Tartelet and it revealed no leaks when runners were restarted a bunch of times.

<img width="1415" alt="Screenshot 2023-07-17 at 14 47 34" src="https://github.com/shapehq/tartelet/assets/830995/f181be41-7c89-471b-bffd-1468a1e0ad60">

**Comment by slashmo** [comment]
I saw the same memory usage spike happening on our Mac Mini M1 (16GB RAM) with Tartelet 0.7.0. We also hit ~ 120GB of memory usage and the runners weren't properly registering anymore, therefore stalling our CI over night. The host runs the latest version of macOS Ventura (`13.6 (22G120)`).

**Comment by phlmn** [comment]
Sorry for the late answer. The funny thing is the issue seems to have disappeared on our system… The only thing I know we have changed is too run only one VM. But I believe we also had memory leaks after the change.

**Comment by xavierLowmiller** [comment]
We're also seeing this on version `0.7.0`:
<img width="429" alt="memory leak" src="https://github.com/shapehq/tartelet/assets/16212751/c50b0dab-a45e-4001-83ed-dc3dbd99ef06">

**Comment by xavierLowmiller** [comment]
> I also ran the Leaks profiling tool on Tartelet and it revealed no leaks when runners were restarted a bunch of times.

Maybe this can be debugged in Xcode 15.1, the [release notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15_1-release-notes#Instruments) mention that the Leaks instrument has had a bug.

**Comment by simonbs** [comment]
I'll make a note to look into this once I get the time but that might not happen until in a few weeks. If anyone else want to take a swing at it in the mean time that would be amazing 🤗

**Comment by weakfl** [comment]
I'm experiencing the same issue with v0.7.1, at least with two VMs.

**Comment by hisaac** [comment]
Throwing my name in the ring here, we're also seeing this issue, running the most recent version of Tartelet. It takes a day or so before it crops up.

**Comment by hisaac** [comment]
So we've got multiple machines running Tartelet. The issue seems to crop up randomly.

One thing I've noticed is that, when I VNC into a host machine that is currently experiencing the problem, I'm presented with a login screen. Once logging in, there are no active Tart VM sessions running, but they immediately start upon logging in.

It appears as if the machines are going to sleep, even though as far as I can tell, all the machines are configured the same, and have the sleep functionality disabled.

**Comment by simonbs** [comment]
I've just checked our three machines which have been maintaining two runners for 42 consecutive days and Tartelet is using around 85 MB of memory on all of them.

I'd love to address this issue but right now I'm unsure where to begin as I'm unable to reproduce it.

**Comment by hisaac** [comment]
Understandable @simonbs. I'd like to keep documenting findings here in case anyone affected has ideas or solutions.

**Comment by hisaac** [comment]
Not sure if either of these will be useful to anyone, but I took a process sample and spindump of Tartelet when I saw it was using high memory on one of our hosts.

- [Process Sample.txt](https://github.com/shapehq/tartelet/files/14125784/Process.Sample.txt)
- [Spindump.txt](https://github.com/shapehq/tartelet/files/14125785/Spindump.txt)

Nothing obvious sticks out to me, but I'm not very experienced reading these.

**Comment by simonbs** [comment]
I wonder if this might be fixed by #97. It ssems like the sort of thing that could cause a memory leak like this.

**Comment by kuhnroyal** [comment]
Yeah, was thinking the same. But haven't seen any big memory even before that.
